bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/ScanAndCompareGarbageCollector.java

```diff
@@ -102,6 +165,68 @@ public void gc(GarbageCleaner garbageCleaner) {
             LOG.warn("Exception when iterating over the metadata {}", e);
         }
     }
-}
 
+    private Set<Long> removeOverReplicatedledgers(Set<Long> bkActiveledgers, final GarbageCleaner garbageCleaner)
+            throws InterruptedException, KeeperException {
+        final Set<Long> overReplicatedLedgers = Sets.newHashSet();
+        final Semaphore semaphore = new Semaphore(MAX_CONCURRENT_ZK_REQUESTS);
+        final CountDownLatch latch = new CountDownLatch(bkActiveledgers.size());
+        for (final Long ledgerId : bkActiveledgers) {
+            try {
+                // check if the ledger is being replicated already by the replication worker
+                if (ZkLedgerUnderreplicationManager.isLedgerBeingReplicated(zk, zkLedgersRootPath, ledgerId)) {
+                    latch.countDown();
+                    continue;
+                }
+                // we try to acquire the underreplicated ledger lock to not let the bookie replicate the ledger that is
+                // already being checked for deletion, since that might change the ledger ensemble to include the
+                // current bookie again and, in that case, we cannot remove the ledger from local storage
+                ZkLedgerUnderreplicationManager.acquireUnderreplicatedLedgerLock(zk, zkLedgersRootPath, ledgerId);
+                semaphore.acquire();
+                ledgerManager.readLedgerMetadata(ledgerId, new GenericCallback<LedgerMetadata>() {
+
+                    @Override
+                    public void operationComplete(int rc, LedgerMetadata ledgerMetadata) {
+                        if (rc == BKException.Code.OK) {
+                            // do not delete a ledger that is not closed, since the ensemble might change again and
+                            // include the current bookie while we are deleting it
+                            if (!ledgerMetadata.isClosed()) {
+                                release();
+                                return;
+                            }
+                            SortedMap<Long, ArrayList<BookieSocketAddress>> ensembles = ledgerMetadata.getEnsembles();
+                            for (ArrayList<BookieSocketAddress> ensemble : ensembles.values()) {
+                                // check if this bookie is supposed to have this ledger
+                                if (ensemble.contains(selfBookieAddress)) {
+                                    release();
+                                    return;
+                                }
+                            }
+                            // this bookie is not supposed to have this ledger, thus we can delete this ledger now
+                            overReplicatedLedgers.add(ledgerId);
```

Comment:
Are we sure it is safe to delete the local ledger fragment here? As a sanity check, I'm trying to determine whether this step is safe.

If a bookie is not in the latest ensemble, then we can't count on its copy of the ledger fragment any longer because the bookie might have crashed permanently. However, we do make an assumption on the number of copies available, so we can't have more permanent failures than the number tolerated and deleting prematurely could lead to data loss. If I'm not  mistaken, we can only delete an over replicated ledger fragment in the case we are sure it already has enough replicas. The checks above only make sure that the ledger isn't closed and the bookie isn't in the latest ensemble.